File: src/model.test.ts
import {importFromUrl, importFromMinUrl, exportAsUrl, Model, exportAsMinUrl} from './model';
import { parseUrl, parseMinUrl, LexModel } from './parser';


const testModel = {
    "modelType": "petriNet",
    "version": "v0",
    "places": {
        "place0": {"offset": 0, "initial": 1, "capacity": 3, "x": 130, "y": 207},
        "place1": {"offset": 1, "x": 395, "y": 299}
    },
    "transitions": {
        "txn0": {"x": 46, "y": 116},
        "txn1": {"x": 227, "y": 112},
        "txn2": {"x": 43, "y": 307},
        "txn3": {"x": 235, "y": 306}
    },
    "arcs": [
        {"source": "txn0", "target": "place0"},
        {"source": "place0", "target": "txn1", "weight": 3},
        {"source": "txn2", "target": "place0", "weight": 3, "inhibit": true},
        {"source": "place0", "target": "txn3", "inhibit": true},
        {"source": "txn3", "target": "place1"}
    ]
} as Model

describe('Model Tests', () => {

  it('should import model from URL', () => {
    const url = '?modelType=testModel&version=1.0&place=place1&offset=10&initial=5&capacity=20&x=100&y=200&transition=trans1&x=300&y=400&source=place1&target=trans1&weight=2&inhibit=true';
    const model: Model = importFromUrl(url);
    expect(model.modelType).toBe('testModel');
    expect(model.version).toBe('1.0');
    expect(model.places['place1']).toEqual({ offset: 10, initial: 5, capacity: 20, x: 100, y: 200 });
    expect(model.transitions['trans1']).toEqual({ x: 300, y: 400 });
    expect(model.arcs[0]).toEqual({ source: 'place1', target: 'trans1', weight: 2, inhibit: true });
  });

  it('should import model from minified URL', () => {
    const minUrl = '?m=testModel&v=1.0&p=place1&o=10&i=5&c=20&x=100&y=200&t=trans1&x=300&y=400&s=place1&e=trans1&w=2&n=1';
    const model: Model = importFromMinUrl(minUrl);
    expect(model.modelType).toBe('testModel');
    expect(model.version).toBe('1.0');
    expect(model.places['place1']).toEqual({ offset: 10, initial: 5, capacity: 20, x: 100, y: 200 });
    expect(model.transitions['trans1']).toEqual({ x: 300, y: 400 });
    expect(model.arcs[0]).toEqual({ source: 'place1', target: 'trans1', weight: 2, inhibit: true });
  });

  it('should parse URL correctly', () => {
    const url = exportAsUrl(testModel)

    const m: LexModel | Error = parseUrl(url);
    if (m instanceof Error) {
        throw m;
    }
    expect(m).toEqual(testModel);
  });

  it('should parse minified URL correctly', () => {
    const minUrl = exportAsMinUrl(testModel)

    const m: LexModel | Error = parseMinUrl(minUrl);
    if (m instanceof Error) {
        throw m;
    }
    expect(m).toEqual(testModel);
  });
});

File: metamodel/model_test.go
package metamodel_test

import (
	. "github.com/pflow-xyz/pflow-app/metamodel"
	"github.com/pflow-xyz/pflow-app/metamodel/cid"
	. "github.com/pflow-xyz/pflow-app/metamodel/token"
	"strings"
	"testing"
)

// TODO: support multiple tokens

var exampleModel = Model{
	ModelType: "PetriNet",
	Version:   "v0",
	Places: map[string]Place{
		"place0": {Offset: 0, Initial: T(1), Capacity: T(3), X: 130, Y: 207},
		"place1": {Offset: 1, Initial: T(0), Capacity: T(0), X: 395, Y: 299},
	},
	Transitions: map[string]Transition{
		"txn0": {X: 46, Y: 116},
		"txn1": {X: 227, Y: 112},
		"txn2": {X: 43, Y: 307},
		"txn3": {X: 235, Y: 306},
	},
	Arrows: []Arrow{
		{Source: "txn0", Target: "place0", Weight: T(1)},
		{Source: "place0", Target: "txn1", Weight: T(3)},
		{Source: "txn2", Target: "place0", Weight: T(3), Inhibit: true},
		{Source: "place0", Target: "txn3", Weight: T(1), Inhibit: true},
		{Source: "txn3", Target: "place1", Weight: T(1)},
	},
}

func TestImportFromJson(t *testing.T) {
	var w strings.Builder
	exampleModel.ToJson(&w)
	json := w.String()
	println(json)
	model := NewModel()
	importedModel, err := model.FromJson(json)
	if err != nil {
		t.Errorf("Expected no error, got %v", err)
	}

	if len(importedModel.Places) != 2 {
		t.Errorf("Expected 2 places, got %v", len(importedModel.Places))
	}
	if len(importedModel.Transitions) != 4 {
		t.Errorf("Expected 4 transitions, got %v", len(importedModel.Transitions))
	}
	if len(importedModel.Arrows) != 5 {
		t.Errorf("Expected 5 arrows, got %v", len(importedModel.Arrows))
	}

	importedCid := cid.NewCid(importedModel).String()
	modelCid := cid.NewCid(model).String()
	if importedCid != modelCid {
		t.Errorf("Expected %v, got %v", modelCid, importedCid)
	}
}
